rust/frontend/src/function.rs

```diff
@@ -0,0 +1,476 @@
+use value::ValueKind;
+use Error;
+use Module;
+use Result;
+use TVMArgValue;
+use TVMRetValue;
+
+lazy_static! {
+    static ref GLOBAL_FUNCTION_NAMES: Mutex<Vec<&'static str>> = list_global_func_names();
+}
+
+fn list_global_func_names() -> Mutex<Vec<&'static str>> {
+    let mut out_size = 0 as c_int;
+    let mut name = ptr::null() as *const c_char;
+    let mut out_array = &mut name as *mut _;
+    check_call!(ts::TVMFuncListGlobalNames(
+        &mut out_size as *mut _,
+        &mut out_array as *mut _
+    ));
+    let list = unsafe { slice::from_raw_parts(out_array, out_size as usize) };
+    let list = list
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_bytes())
+        .map(|bs| str::from_utf8(bs).unwrap())
+        .collect();
+    Mutex::new(list)
+}
+
+/// Returns a registered TVM function by name.
+pub fn get_global_func(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+    let name = CString::new(name).unwrap();
+    let mut handle = ptr::null_mut() as ts::TVMFunctionHandle;
+    check_call!(ts::TVMFuncGetGlobal(
+        name.as_ptr() as *const c_char,
+        &mut handle as *mut _
+    ));
+    if !(handle.is_null()) {
+        mem::forget(name);
+        return Some(Function::new(handle, is_global, false));
+    } else {
+        if allow_missing {
+            return None;
+        } else {
+            panic!("Cannot find global function {:?}", name);
+        }
+    }
+}
+
+/// Wrapper around TVM function handle.
+#[derive(Debug, Hash)]
+pub struct Function {
+    handle: ts::TVMFunctionHandle,
+    is_global: bool,
+    is_released: bool,
+}
+
+impl Function {
+    pub(crate) fn new(handle: ts::TVMFunctionHandle, is_global: bool, is_released: bool) -> Self {
+        Function {
+            handle: handle,
+            is_global: is_global,
+            is_released: is_released,
+        }
+    }
+
+    /// For a given function, it returns a function by name.
+    pub fn get_function(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+        let name = CString::new(name).unwrap();
+        GLOBAL_FUNCTION_NAMES
+            .lock()
+            .unwrap()
+            .iter()
+            .find(|&&s| s == name.to_str().expect("Invalid UTF-8 name"))
+            .map(|nm| get_global_func(nm, is_global, allow_missing).unwrap())
+    }
+
+    /// Returns the underlying TVM function handle.
+    pub fn as_handle(&self) -> ts::TVMFunctionHandle {
+        self.handle
+    }
+
+    pub fn is_global(&self) -> bool {
+        self.is_global
+    }
+
+    pub fn is_released(&self) -> bool {
+        self.is_released
+    }
+}
+
+impl Clone for Function {
+    fn clone(&self) -> Function {
+        if !self.is_released {
+            Self {
+                handle: self.handle,
+                is_global: true,
+                is_released: false,
+            }
+        } else {
+            panic!("Released function cannot be cloned!");
```

Comment:
I would actually just allow this and error when the function is used. After all, passing around null pointers isn't the problem: dereferencing them is.

Reply:
Initially I didn't put `!self.is_released` restriction on `clone`, but then I got segfaults when multiple functions need to co-exist and `function::Builder` needs to own each function. After that, the only option in `else` is to `self.clone()` which seems weird to me!

Reply:
that doesn't sound quite right. if you're not seeing panics, then this function isn't actually doing anything different from `#[derive(Clone)]`.

Reply:
The problem is for non-global functions such as `load_params` which comes from a graph runtime module, as soon as I use it once, it's released and if I get another non-global function like `set_input` from graph runtime, it segfaults. So in clone I have to make that distinction and temporarily set non-global as `is_global = true` in clone which is not ideal I know!

Reply:
yeah that's actually really confusing. do you think there's a way around this? I've never seen `clone` used as a way to designate something as non-released. Perhaps make an explicit `to_global(self) -> Self` method?

Reply:
there are three TODOs now: this, refactoring builder to use a derive, and a macro for constructing the builder. Let's not lose track of these and make an issue on #1601 for relatively-soon consideration.

Reply:
(#2303)[https://github.com/dmlc/tvm/issues/2303], (#2304)[https://github.com/dmlc/tvm/issues/2304], (#2305)[https://github.com/dmlc/tvm/issues/2305]

Reply:
wow nice, actual issues! thanks a lot!

Reply:
The current fix I came up with is to add `is_cloned` to `Function`. Then, turning it one on, on `clone()` would hint `drop` not dropping from the frontend, no matter `is_global = false` (for example, in `load_params`, `set_input` etc.)

Reply:
I don't quite understand the prose, but the idea sounds reasonable 👍

Reply:
Well, we don't need to `panic!` and clone just creates a fresh new function. Then clone and drops will be more natural like

```
impl Clone for Function {
    fn clone(&self) -> Function {
        if !self.is_released && !self.is_cloned {
            Self {
                handle: self.handle,
                is_global: self.is_global,
                is_released: self.is_released,
                is_cloned: true,
            }
        } else {
            Function::new(self.handle, self.is_global, self.is_released)
        }
    }
}

impl Drop for Function {
    fn drop(&mut self) {
        if !self.is_released && !self.is_global && !self.is_cloned {
            check_call!(ts::TVMFuncFree(self.handle));
            self.is_released = true;
        }
    }
}
```